### PR TITLE
Update semantics description in documentation

### PR DESCRIPTION
**Pull request by kathrynmohror**
### Description
Our description of UnifyFS semantics in our documentation is woefully out of date. This PR fleshes out the description and updates it to be current with the implementation.

### Motivation and Context
Reorganize section, add more detail, clarify semantics because sorely out of date.

### How Has This Been Tested?
Not tested

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Testing (addition of new tests or update to current tests)
- [x ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the UnifyFS code style requirements.
- [x ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted.

### CODE REVIEW

File: docs/assumptions.rst
Comment (MichaelBrim) [inline docs/assumptions.rst:131]: I think the paragraph is mostly right in that concurrent writes to the same offset/region will result in undefined behavior, but perhaps it's not as clear as it could be. The problem with the current discussion is we're overloading "synchronization operations" to mean two things that are not equivalent. One type is local synchronization (i.e., flush extents to the server), the other type is global synchronization across app processes (which must be provided by something like MPI since Unify does not provide anything for that). For a consistent distributed sequential ordering of conflicting writes, there must be global sync ops between each of the local sync ops.

File: docs/assumptions.rst
Comment (MichaelBrim) [inline docs/assumptions.rst:146]: I think we need a separate section on failures, because it really depends on where the failure occurs, regardless of lamination state. If a node goes down, it takes the server with it, and all data from the node is lost. If the application has a terminal failure, and the server is still running, another local or remote client can still access the data. If the server encounters a terminal failure, all data on the node is lost.

File: docs/assumptions.rst
Comment (MichaelBrim) [inline docs/assumptions.rst:186]: yes, see above comment

File: docs/assumptions.rst
Comment (MichaelBrim) [inline docs/assumptions.rst:237]: "libraries libraries"

File: docs/assumptions.rst
Comment (MichaelBrim) [inline docs/assumptions.rst:250]: no, see above comment

**Review by MichaelBrim — CHANGES_REQUESTED** [review]

File: docs/assumptions.rst
Comment (kathrynmohror) [reply docs/assumptions.rst:147]: Thanks @MichaelBrim. I added a new section on failure behavior

File: docs/assumptions.rst
Comment (kathrynmohror) [reply docs/assumptions.rst:114]: I agree. I misread our design document and got it in my head that the library could read any of the local extent trees (from any process). I fixed it to say that sync ops are needed for this case.

File: docs/assumptions.rst
Comment (kathrynmohror) [reply docs/assumptions.rst:132]: I added a discussion of local and global synchronization near the top of this section. I think it helps, thanks for suggesting it. 

I still wonder about whether the most recent update will always be returned, even if correct global synchronization is used. Perhaps, though, this was a problem with MDHIM and is fixed with our new approach? I think with MDHIM we ended up with multiple key,value pairs for the same offset and we didn't know which one was most recent so one would be returned (but perhaps not the most recent). Perhaps now with our new approach, there is just one location where data for an offset is stored and when a sync operation is applied, that location is updated with the most recent data. Is that right?

File: docs/assumptions.rst
Comment (adammoody) [inline docs/assumptions.rst:32]: maybe also call out "overlapped writes" in the list of things not to do in bulk synchronous mode?

File: docs/assumptions.rst
Comment (adammoody) [inline docs/assumptions.rst:68]:
```suggestion
the bulk synchronous I/O pattern, and it enables UnifyFS to provide better
```

File: docs/assumptions.rst
Comment (adammoody) [reply docs/assumptions.rst:152]: For "global sync", the word "global" is a bit strong.  What really matters is "inter-process sync", just the subset of processes that care about some shared offset, which might just be two procs out of many.  It's more like "file system synchronization" and "process synchronization".

Comment (adammoody) [reply docs/assumptions.rst:152]: That's right, we don't have the problem with two key/val pairs existing for a single file offset anymore.  A single owner server is responsible for updating the main extent tree for a file.  Old entries are replaced by new entries.

Another thing regarding "undefined".  From the user's point of view, if two procs write to the same region, the result could be some strange mix of bytes from the two, even if each process issued one write.  For example, consider that two procs each do a single large write of say 1GB to the same region of the file.  Internally, each single large write may be broken into a bunch of smaller chunks/extents, which are sent to the owner server maintaining the extent tree in numerous rpc calls.  Those then get played on the server in random order.  The result is that we might get a mix of data from the two procs for that region.  It's not as if the single large write from one of the two procs wins and the other loses.  Our current implementation might actually do better than what I'm describing, but I think it'd be good to stress the "undefined" bit until users complain since it gives us some flexibility.

File: docs/assumptions.rst
Comment (adammoody) [inline docs/assumptions.rst:199]: Maybe tweak this to highlight that "well-behaved" apps can benefit from the extreme write buffering optimization.
```suggestion
For applications in which processes do not need to access file data modified
by other processes before lamination,
```

File: docs/assumptions.rst
Comment (adammoody) [reply docs/assumptions.rst:32]: Just for the team... I wouldn't change the text here, but technically we should be able to handle random and interleaved writes just fine.

We could also handle small writes efficiently.  Chunk space utilization and our extent tree can run into problems.  For extreme cases, we have it on our roadmap that we could also support a stripe-based storage model like Lustre rather than using a log-based I/O model.  The stripe-based method eliminates the need for extents and it can fully pack the storage, though it means we need to transfer data bytes on sync operations rather than extent info.  The write data could still be buffered locally until the sync/laminate, though.

Small local reads should also be decent in our current model, but lots of small remote reads will be bad.

Comment (adammoody) [reply docs/assumptions.rst:32]: On when to switch to a striped-based method, there are a few interesting options:
1) allow user to set that as a config
2) use log-I/O during write, reformat to a stripe-based layout at laminate
3) track whether the extents for a file consume more space than the data as the file is being written, and switch layouts on the fly

File: docs/assumptions.rst
Comment (MichaelBrim) [inline docs/assumptions.rst:117]: remove period before closing paren

File: docs/assumptions.rst
Comment (MichaelBrim) [reply docs/assumptions.rst:199]: +1

File: docs/assumptions.rst
Comment (kathrynmohror) [reply docs/assumptions.rst]: I made some updates to introduce the concept of overlapping writes and to drive home that the result in that case is undefined.

**Comment by adammoody** [comment]
@MichaelBrim , the merge button is blocked pending a change request from you.  Is that good to go now?  If so, do you see where you can sign off again?

**Review by MichaelBrim — APPROVED** [review]

**Comment by MichaelBrim** [comment]
@adammoody should be good to merge now

**Pull request merged**
